core/offchain.go

```diff
@@ -94,25 +96,24 @@ func (bc *BlockChain) CommitOffChainData(
 	//	}
 	//}
 
-	newEpoch := new(big.Int).Add(header.Epoch(), common.Big1)
+	nextBlockEpoch, err := bc.getNextBlockEpoch(header)
+	if err != nil {
+		return NonStatTy, err
+	}
+
 	// Shard State and Validator Update
 	if isNewEpoch {
 		// Write shard state for the new epoch
-		newShardState, err := bc.WriteShardStateBytes(batch, newEpoch, header.ShardState())
+		_, err := bc.WriteShardStateBytes(batch, nextBlockEpoch, header.ShardState())
 		if err != nil {
 			header.Logger(utils.Logger()).Warn().Err(err).Msg("cannot store shard state")
 			return NonStatTy, err
 		}
-
-		if err == nil && newShardState.Epoch != nil && bc.chainConfig.IsStaking(newShardState.Epoch) {
-			// After staking, the epoch will be decided by the epoch in the shard state.
-			newEpoch = new(big.Int).Set(newShardState.Epoch)
-		}
 	}
 
 	// Do bookkeeping for new staking txns
 	newVals, err := bc.UpdateStakingMetaData(
-		batch, block, state, epoch, newEpoch,
+		batch, block, state, epoch, nextBlockEpoch,
```

Comment:
so, this update was wrong before? as it would right write snapshot for newly created validator even before NewEpoch?

Reply:
Right, it was duplicate snapshotted, But since those validator will be snapshotted again, the wrong snapshot will be overwritten. So it wasn't causing any issue, but main potential issue is the WriteShardState part.